engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/formatter/ConfluentAvroMessageReader.scala

```diff
@@ -1,94 +1,64 @@
 
 /**
-  * This class is mainly copy-paste of Confluent's AvroMessageReader but with better constructor handling
-  * both passing schemaRegistryClient and keySeparator.
-  *
   * @param schemaRegistryClient schema registry client
   * @param topic topic
-  * @param keySeparator key separator
   */
-private[confluent] class ConfluentAvroMessageReader(schemaRegistryClient: SchemaRegistryClient,
-                                                    topic: String,
-                                                    useStringForKey: Boolean,
-                                                    keySeparator: String)
+private[confluent] class ConfluentAvroMessageReader(schemaRegistryClient: SchemaRegistryClient, topic: String)
   extends AbstractKafkaAvroSerializer with DatumReaderWriterMixin {
 
   schemaRegistry = schemaRegistryClient
 
-  private val keySubject = ConfluentUtils.keySubject(topic)
-
-  private val valueSubject = ConfluentUtils.valueSubject(topic)
-
   private val decoderFactory = DecoderFactory.get
 
-  // TODO: This implementation won't handle separator escaping
-  def readMessage(str: String, keySchema: Option[Schema], valueSchema: Option[Schema]): ConsumerRecord[Array[Byte], Array[Byte]] = {
+  def schemaById(schemaId: Int): Schema = {
+    val parsedSchema = schemaRegistryClient.getSchemaById(schemaId)
+    ConfluentUtils.extractSchema(parsedSchema)
+  }
+
+  def readJson[T: ClassTag](jsonObj: Json, schemaOpt: Option[Schema], subject: String): Array[Byte] = {
     try {
-      val keyIndex = str.indexOf(keySeparator)
-      if (keyIndex < 0) {
-        throw new SerializationException("No key found in line " + str)
-      } else {
-        val keyString = str.substring(0, keyIndex)
-        val valueString = if (keyIndex + 1 > str.length) "" else str.substring(keyIndex + 1)
-        keySchema match {
-          case None =>
-            val value = jsonToAvro(valueString, valueSchema.get)
-            val serializedValue = serializeImpl(valueSubject, value, new AvroSchema(valueSchema.get))
-            new ConsumerRecord(topic, 0, 0L, Array[Byte](), serializedValue)
-          case Some(_) =>
-            // default scenario when key and value is provided
-            if (keyString.length == 0) {
-              // key schema is provided but there is no content
-              throw new SerializationException("No key content found in line " + str)
-            }
-            val serializedKey = if (useStringForKey) {
-              keyString.getBytes(StandardCharsets.UTF_8)
-            } else {
-              val key = jsonToAvro(keyString, keySchema.get)
-              serializeImpl(keySubject, key, new AvroSchema(keySchema.get))
-            }
-            val value = jsonToAvro(valueString, valueSchema.get)
-            val serializedValue = serializeImpl(valueSubject, value, new AvroSchema(valueSchema.get))
-            new ConsumerRecord(topic, 0, 0L, serializedKey, serializedValue)
-        }
+      schemaOpt match {
```

Comment:
Returning `emptyByteArray` is not too much intuitive. Can we make this class easyly handle only defined schema and handle all edge cases in one places - in formatter?

Reply:
Fixed